src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
+        ret = -1;
+    }
 
-    WOLFSSL_LEAVE("wolfSSL_DH_compute_key", ret);
+#endif /* !HAVE_FIPS || FIPS_VERION > 2 */
 
     return ret;
 }
 
+#endif
 
-#if defined(OPENSSL_ALL) || \
-    defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000L
-int wolfSSL_DH_set_length(WOLFSSL_DH *dh, long len)
-{
-    WOLFSSL_ENTER("wolfSSL_DH_set_length");
-
-    /* len is checked at generation */
-    if (dh == NULL) {
-        WOLFSSL_MSG("Bad function arguments");
-        return 0;
-    }
+/*
+ * DH PEM APIs
+ */
 
-    dh->length = (int)len;
-    return 1;
-}
+#if defined(HAVE_LIGHTY) || defined(HAVE_STUNNEL) \
+    || defined(WOLFSSL_MYSQL_COMPATIBLE) || defined(OPENSSL_EXTRA)
 
-/* ownership of p,q,and g get taken over by "dh" on success and should be free'd
- * with a call to wolfSSL_DH_free -- not individually.
+#if !defined(NO_BIO) || !defined(NO_FILESYSTEM)
+/* Create a DH key by reading the PEM encoded data from the BIO.
+ *
+ * DH parameters are public data and are not expected to be encrypted.
  *
- * returns 1 on success
+ * @param [in]      bio   BIO object to read from.
+ * @param [in, out] dh    DH key with parameters if not NULL. When pointer to
+ *                        NULL, a new DH key is created.
+ * @param [in]      cb    Password callback when PEM encrypted. Not used.
+ * @param [in]      pass  NUL terminated string for passphrase when PEM
+ *                        encrypted. Not used.
+ * @return  DH key on success.
+ * @return  NULL on failure.
  */
-int wolfSSL_DH_set0_pqg(WOLFSSL_DH *dh, WOLFSSL_BIGNUM *p,
-    WOLFSSL_BIGNUM *q, WOLFSSL_BIGNUM *g)
+static WOLFSSL_DH *wolfssl_dhparams_read_pem(WOLFSSL_DH **out,
+    unsigned char* pem, int pemSz, int memAlloced, wc_pem_password_cb *cb,
+    void *pass)
 {
-    int ret;
-    WOLFSSL_ENTER("wolfSSL_DH_set0_pqg");
+    WOLFSSL_DH* localDh = NULL;
+    DerBuffer *der = NULL;
+    int err = 0;
 
-    /* q can be NULL */
-    if (dh == NULL || p == NULL || g == NULL) {
-        WOLFSSL_MSG("Bad function arguments");
-        return 0;
-    }
+    /* Parameters are public data and are not encrypted. */
+    (void)cb;
+    (void)pass;
 
-    /* free existing internal DH structure and recreate with new p / g */
-    if (dh->inSet) {
-#ifndef HAVE_SELFTEST
-        ret = wc_FreeDhKey((DhKey*)dh->internal);
-        if (ret != 0) {
-            WOLFSSL_MSG("Unable to free internal DH key");
-            return 0;
+    /* Convert PEM to DER assuming DH Parameter format. */
+    if ((!err) && (PemToDer(pem, pemSz, DH_PARAM_TYPE, &der, NULL, NULL,
+            NULL) < 0)) {
+        /* Convert PEM to DER assuming X9.42 DH Parameter format. */
+        if (PemToDer(pem, pemSz, X942_PARAM_TYPE, &der, NULL, NULL, NULL)
+                != 0) {
+            err = 1;
         }
-#else
-        /* Selftest code has this API with a void return type */
-        wc_FreeDhKey((DhKey*)dh->internal);
-#endif
+    }
+    if (memAlloced) {
```

Comment:
To me, this is confusing behavior. I wouldn't expect this function to free memory that I, the caller, allocated and passed as an input parameter. A more intuitive interface would be to remove this and require the caller to free their own memory.

Reply:
Reduces high water mark on memory allocated.